### PR TITLE
added mainnet fork arbitrum test

### PR DESCRIPTION
**Pull request by crisgarner**
test with:

``` forge test --match-path ./test/arbitrum/JPEGzMainnetOracle.t.sol --fork-url INFURA_URL```

### CODE REVIEW

File: test/arbitrum/JPEGzMainnetOracle.t.sol [hunk 1]
```diff
@@ -0,0 +1,505 @@
...
+  AggregatorV3Interface jpegzAggregator =
+    AggregatorV3Interface(0x8D0e319eBAA8DF32e088e469062F85abF2eBe599);
+  AggregatorV3Interface ethAggregator =
+    AggregatorV3Interface(0x639Fe6ab55C921f74e7fac1ee960C0B6293ba612);
+  ChainlinkOracle jpegzOracle =
+    new ChainlinkOracle(address(jpegzAggregator), address(this));
+  ChainlinkOracle ethOracle =
+    new ChainlinkOracle(address(ethAggregator), address(this));
+  address weth = 0x82aF49447D8a07e3bd95BD0d56f35241523fBab1;
+  address user = address(0x1);
```
Comment (voith) [inline test/arbitrum/JPEGzMainnetOracle.t.sol:30]: Please change this address as this is a precompile address. 
I use something like address(0x51);

Comment (voith) [inline test/arbitrum/JPEGzMainnetOracle.t.sol:30]: Please change this address as this is a precompile address. 
I use something like `address(0x51)`.

File: test/arbitrum/JPEGzMainnetOracle.t.sol [hunk 2]
```diff
@@ -0,0 +1,505 @@
...
+  address user2 = address(0x2);
+
+  //// Params
+  uint256 divisor = 1000000000;
+  uint256 ratio = 150;
+  uint256 burnFee = 150;
+  uint256 mintFee = 150;
+  uint256 liquidationPenalty = 20;
+  address treasury = address(0x3);
```
Comment (voith) [inline test/arbitrum/JPEGzMainnetOracle.t.sol:39]: same as above

File: test/arbitrum/JPEGzMainnetOracle.t.sol [hunk 3]
```diff
@@ -0,0 +1,505 @@
...
+    assertEq(address(ethOracle), address(ethVault.collateralPriceOracle()));
+    assertEq(address(ethOracle), address(ethVault.ETHPriceOracle()));
+    assertEq(treasury, ethVault.treasury());
+    assertEq(0 ether, ethVault.minimumTCAP());
+  }
+
+  function testConstructor_ShouldRevert_WhenBurnFeeIsHigh(uint256 _burnFee)
+    public
+  {
+    if (_burnFee > 1000) {
```
Comment (voith) [inline test/arbitrum/JPEGzMainnetOracle.t.sol:80]: Forge has a trick for fuzzing: `vm.assume(_burnFee > 1000)`.
With this trick, you should be able to get rid of the if block.

File: test/arbitrum/JPEGzMainnetOracle.t.sol [hunk 4]
```diff
@@ -0,0 +1,505 @@
...
+    assertEq(burnFee, ethVault.burnFee());
+    assertEq(burnFee, ethVault.mintFee());
+    assertEq(liquidationPenalty, ethVault.liquidationPenalty());
+    assertEq(address(jpegzOracle), address(ethVault.tcapOracle()));
+    assertEq(address(jpegz), address(ethVault.TCAPToken()));
+    assertEq(address(weth), address(ethVault.collateralContract()));
+    assertEq(address(ethOracle), address(ethVault.collateralPriceOracle()));
+    assertEq(address(ethOracle), address(ethVault.ETHPriceOracle()));
+    assertEq(treasury, ethVault.treasury());
+    assertEq(0 ether, ethVault.minimumTCAP());
```
Comment (voith) [inline test/arbitrum/JPEGzMainnetOracle.t.sol:74]: wish we had refactored stuff like this which says TCAP :(

File: test/arbitrum/JPEGzMainnetOracle.t.sol [hunk 5]
```diff
@@ -0,0 +1,505 @@
...
+      jpegz,
+      address(weth),
+      address(ethOracle),
+      address(ethOracle),
+      treasury,
+      0
+    );
+
+    if (!(_burnFee > 1000)) {
+      assertEq(_burnFee, vault.burnFee());
```
Comment (voith) [inline test/arbitrum/JPEGzMainnetOracle.t.sol:101]: This should be a separate test. The test name is `ShouldRevert_WhenBurnFeeIsHigh`. 
Please split these into two separate tests.

File: test/arbitrum/JPEGzMainnetOracle.t.sol [hunk 6]
```diff
@@ -0,0 +1,505 @@
...
+    }
+  }
+
+  function testConstructor_ShouldRevert_WhenMintFeeIsHigh(uint256 _mintFee)
+    public
+  {
+    vm.assume(_mintFee > 1000);
```
Comment (voith) [inline test/arbitrum/JPEGzMainnetOracle.t.sol:108]: seems like you already know about `vm.assume` :)

File: test/arbitrum/JPEGzMainnetOracle.t.sol [hunk 7]
```diff
@@ -0,0 +1,505 @@
...
+      treasury,
+      0
+    );
+  }
+
+  function testConstructor_ShouldRevert_WhenLiquidationPenaltyIsHigh(
+    uint256 _liquidationPenalty,
+    uint256 _ratio
+  ) public {
+    if (_liquidationPenalty + 100 < _liquidationPenalty) {
```
Comment (voith) [inline test/arbitrum/JPEGzMainnetOracle.t.sol:132]: Can we use `vm.assume(_liquidationPenalty + 100) >= _ratio)`?

File: test/arbitrum/JPEGzMainnetOracle.t.sol [hunk 8]
```diff
@@ -0,0 +1,505 @@
...
+      jpegz,
+      address(weth),
+      address(ethOracle),
+      address(ethOracle),
+      treasury,
+      0
+    );
+
+    if (!((_liquidationPenalty + 100) >= _ratio)) {
+      assertEq(_liquidationPenalty, vault.liquidationPenalty());
```
Comment (voith) [inline test/arbitrum/JPEGzMainnetOracle.t.sol:159]: Please split this as a separate as this test is different from its name i.e `ShouldRevert_WhenLiquidationPenaltyIsHigh`.

File: test/arbitrum/JPEGzMainnetOracle.t.sol [hunk 9]
```diff
@@ -0,0 +1,505 @@
...
+    }
+  }
+
+  function testSetLiquidationPenalty_ShouldUpdateValue(
+    uint256 _liquidationPenalty
+  ) public {
+    if (_liquidationPenalty + 100 < _liquidationPenalty) {
```
Comment (voith) [inline test/arbitrum/JPEGzMainnetOracle.t.sol:166]: can we use `vm.assume` here?

File: test/arbitrum/JPEGzMainnetOracle.t.sol [hunk 10]
```diff
@@ -0,0 +1,505 @@
...
+      vm.expectRevert(
+        "VaultHandler::setLiquidationPenalty: liquidation penalty too high"
+      );
+    }
+
+    orchestrator.setLiquidationPenalty(ethVault, _liquidationPenalty);
+  }
+
+  function testShouldUpdateRatio(uint256 _ratio) public {
+    if (_ratio < 100) {
```
Comment (voith) [inline test/arbitrum/JPEGzMainnetOracle.t.sol:180]: can we do `vm.assume((_ratio < 100) &&(ethVault.liquidationPenalty() + 100 >= _ratio) )`?

File: test/arbitrum/JPEGzMainnetOracle.t.sol [hunk 11]
```diff
@@ -0,0 +1,505 @@
...
+    orchestrator.executeTransaction(
+      address(ethVault),
+      0,
+      "setMinimumTCAP(uint256)",
+      abi.encode(_minimumTCAP)
+    );
+    assertEq(ethVault.minimumTCAP(), _minimumTCAP);
+  }
+
+  function testMint_ShouldCreateTCAP_WhenFeeIsPaid() public {
```
Comment (voith) [inline test/arbitrum/JPEGzMainnetOracle.t.sol:205]:
```suggestion
  function testMint_ShouldCreateJPEGZ_WhenFeeIsPaid() public {
```

File: test/arbitrum/JPEGzMainnetOracle.t.sol [hunk 12]
```diff
@@ -0,0 +1,505 @@
...
+      vm.expectRevert("VaultHandler::setRatio: ratio lower than MIN_RATIO");
+    } else {
+      if (ethVault.liquidationPenalty() + 100 >= _ratio) {
+        vm.expectRevert("VaultHandler::setRatio: liquidation penalty too high");
+      }
+    }
+    orchestrator.setRatio(ethVault, _ratio);
+  }
+
+  function testSetMinimumTCAP_ShouldUpdateValue(uint256 _minimumTCAP) public {
```
Comment (voith) [inline test/arbitrum/JPEGzMainnetOracle.t.sol:190]:
```suggestion
  function testSetMinimumJPEGZ_ShouldUpdateValue(uint256 _minimumTCAP) public {
```

File: test/arbitrum/JPEGzMainnetOracle.t.sol [hunk 13]
```diff
@@ -0,0 +1,505 @@
...
+    ethVault.addCollateralETH{value: 10 ether}();
+    ethVault.mint{value: fee}(1 ether);
+
+    (uint256 id, uint256 collateral, uint256 debt, address owner) = ethVault
+      .vaults(1);
+    assertEq(id, 1);
+    assertEq(collateral, 10 ether);
+    assertEq(debt, 1 ether);
+    assertEq(owner, user);
+    assertEq(address(treasury).balance, fee);
```
Comment (voith) [inline test/arbitrum/JPEGzMainnetOracle.t.sol:219]: Just to be sure that our calculations with fees are correct, can you please check that this actually 1.5% of 1 ether?
i.e 
```solidity
unit256 _fee = ((15/10) * 1 ether * JPEGZ_PRICE())/(100 * ETH_PRICE());
assertEq(address(treasury).balance, _fee);
```
using 15/10 as we know that is equal 1.5. I just want to make sure that we've set `uint256 mintFee = 150;` correctly above.

File: test/arbitrum/JPEGzMainnetOracle.t.sol [hunk 14]
```diff
@@ -0,0 +1,505 @@
...
+    ethVault.mint{value: fee}(9 ether);
+
+    // asserts
+    (, , debt, ) = ethVault.vaults(1);
+    assertEq(debt, 30 ether);
+  }
+
+  function testBurnTCAP_ShouldBurn_WhenFeeIsPaid(uint96 _tcapAmount) public {
+    // checks
+    if (_tcapAmount < 1 ether) {
```
Comment (voith) [inline test/arbitrum/JPEGzMainnetOracle.t.sol:303]: use `vm.assume`

File: test/arbitrum/JPEGzMainnetOracle.t.sol [hunk 15]
```diff
@@ -0,0 +1,505 @@
...
+    orchestrator.executeTransaction(
+      address(ethVault),
+      0,
+      "setMinimumTCAP(uint256)",
+      abi.encode(1 ether)
+    );
+    vm.startPrank(user);
+    uint256 requiredCollateral = ethVault.requiredCollateral(_tcapAmount);
+    uint256 mfee = ethVault.getBurnFee(_tcapAmount);
+    uint256 bfee = ethVault.getBurnFee(_tcapAmount);
```
Comment (voith) [inline test/arbitrum/JPEGzMainnetOracle.t.sol:318]: Can you test the value of bfee like the way I've requested you to test the mintFee?

File: test/arbitrum/JPEGzMainnetOracle.t.sol [hunk 16]
```diff
@@ -0,0 +1,505 @@
...
+
+    //assert
+    assertEq(oldFee, ethVault.burnFee());
+  }
+
+  function testSetBurnFee_ShouldAllowDecimals_WhenValueBelowMax(uint256 _newFee)
+    public
+  {
+    //setUp
+    if (_newFee > 1000) {
```
Comment (voith) [inline test/arbitrum/JPEGzMainnetOracle.t.sol:388]: use `vm.assume`

File: test/arbitrum/JPEGzMainnetOracle.t.sol [hunk 17]
```diff
@@ -0,0 +1,505 @@
...
+    //assert
+    assertEq(ethVault.mintFee(), _newFee);
+  }
+
+  function testGetMintFee_ShouldCalculateCorrectValue(
+    uint8 _mintFee,
+    uint96 _amount
+  ) public {
+    //setUp
+    if (_mintFee > 1000) {
```
Comment (voith) [inline test/arbitrum/JPEGzMainnetOracle.t.sol:450]: use `vm.assume`

**Review by voith — COMMENTED** [review]
Please change the addresses that use precompile addresses. 
Also, I've requested some cleanup.

File: test/arbitrum/JPEGzMainnetOracle.t.sol (hunk 4 above)
Comment (crisgarner) [reply test/arbitrum/JPEGzMainnetOracle.t.sol:74]: I agree, especially when we try to make tests it throws an error that JPGEZ != TCAP. so in the end we end using TCAP contract

File: test/arbitrum/JPEGzMainnetOracle.t.sol (hunk 5 above)
Comment (crisgarner) [reply test/arbitrum/JPEGzMainnetOracle.t.sol:101]: Fixed, this is already tested on the testConstructor_ShouldSetParams_WhenInitialized

File: test/arbitrum/JPEGzMainnetOracle.t.sol (hunk 6 above)
Comment (crisgarner) [reply test/arbitrum/JPEGzMainnetOracle.t.sol:108]: yeah, seems like I had some old test and forgot to update them

File: test/arbitrum/JPEGzMainnetOracle.t.sol (hunk 13 above)
Comment (crisgarner) [reply test/arbitrum/JPEGzMainnetOracle.t.sol:219]: added these lines:

```
uint256 _fee = ((15 / 10) * 1 ether * ethVault.getOraclePrice(jpegzOracle)) / (ethVault.oracleDigits() * 10 * ethVault.getOraclePrice(ethOracle));
assertEq(address(treasury).balance, _fee);

```

**Comment by crisgarner** [comment]
Fixed the observations @voith. Fixed the test also. In order to test the fork now you need to set up the .env variable `FOUNDRY_PROFILE=fork` and disable it for normal testing.

fork is now tested with

`forge test --fork-url RPC`

**Review by voith — APPROVED** [review]
LGTM 👍

**Pull request merged**